### PR TITLE
Misc enhancements regarding issues

### PR DESCRIPTION
**Pull request by zhongh**

### CODE REVIEW

**Comment by zhongh** [comment]
image
image

**Comment by zednis** [comment]
The first legend seems to indicate that a color is used with age to indicate some information, yet I do not see any color used with age in the screenshots.  Are we intending to add color to the age values in the UI and if so how will this information be encoded in the elasticsearch JSON?

File: ingest/mappings/mineral-observation.json [hunk 1]
```diff
@@ -148,6 +142,25 @@
             "type":"string"
           }
         }
+      },
+      "mineral-specimen":{
```
Comment (zednis) [inline ingest/mappings/mineral-observation.json:146]: :+1: 

will information on how the age was determined (G, B, Y, O, R) be encoded in a property of the specimen?

**Comment by zhongh** [comment]
With previous commiting taking ages as numerical types, the range facets of min/max ages are working:
image

**Comment by zednis** [comment]
I think the min/max age range facets are a good first start before we add the geologic age information.

File: ingest/mappings/mineral-observation.json (hunk 1 above)
Comment (zhongh) [inline ingest/mappings/mineral-observation.json:146]: Currently it's only recorded in the specimen as "G", "B", ... This could be the next step?

**Comment by zednis** [comment]
fixes #19

**Comment by zednis** [comment]
fixes #15

**Comment by zhongh** [comment]
fixes #16

File: ingest/mappings/mineral-observation.json [hunk 2]
```diff
@@ -148,6 +142,25 @@
...
+        "properties":{
+          "max-age":{
+            "type":"double"
+          },
+          "min-age":{
+            "type":"double"
+          },
+          "label-1":{
```
Comment (zednis) [inline ingest/mappings/mineral-observation.json:157]: @xgmachina @kolans do you know if there is any meaning behind label-1, label-2, label-3 in the RRUFF spreadsheets?  What is the difference between these three columns?

**Comment by zhongh** [comment]
fixes #11

File: ingest/mappings/mineral-observation.json (hunk 1 above)
Comment (zednis) [inline ingest/mappings/mineral-observation.json:146]: Let's make the description of the age determination process an object with a label and description.

example:

```
"mineral-specimen" : {
  "age-determination-process": {
    "label": "G",
    "description": "This mineral is directly dated." 
  }
}
```

File: html/browser.html [hunk 3]
```diff
@@ -29,17 +29,31 @@
...
-                    <a href="http://www.google.com/maps/place/{{locality.decimal-degree}}" target="_blank"> (View in Google Map)</a>
+                    <a href="http://www.google.com/maps/place/{{locality.coordinates}}" target="_blank"> (View in Google Map)</a>
                 </div>
 
-                {{#if locality.min-age}}<div><strong>Min Age:</strong> {{locality.min-age}}</div>{{/if}}
-                {{#if locality.max-age}}<div><strong>Max Age:</strong> {{locality.max-age}}</div>{{/if}}
+                {{#if mineral-specimen.min-age}}
+                <div>
+                    <strong>Min Age:</strong> {{mineral-specimen.min-age}} {{mineral-specimen.age-unit}}
+                    {{#if mineral-specimen.label-1}}<sup>{{mineral-specimen.label-1}}</sup>{{/if}}
```
Comment (zednis) [inline html/browser.html:38]: So mineral-specimen.label-1 is for the age?

**Comment by zednis** [comment]
Next steps:
- [ ] determine what if mineral-specimen.label-1, mineral-specimen.label-2, mineral-specimen.label-3 mean.  Rename the properties properties to something meaningful (ask @xgmachina @kolans)
- [x] add age unit property
- [ ] add age-determination-process object (not final name)

**Comment by zhongh** [comment]
Today Sophie and me had a pretty good conversation regarding some geoscience-specific questions. Here are what are related to the remaining tasks of this branch:
1. We are going to add an object under mineral-observation.mineral-specimen recording the information on how the age is determined. An initial suggestion is to call it "**age-determination-process**", but maybe a slightly different "**age-determination-info**" could be a more precise name, as a "process" could be a way bigger concept.
2. For now, the first 3 columns of an evolution table are recorded in objects **mineral-specimen.label-1**, **mineral-specimen.label-2**, and **mineral-specimen.label-3**. As far as we know, they record these information and should be renamed as suggested below:
   - **mineral-specimen.label-1** either has value "MD" or no value, where "MD" remains unknown (could be "mindat", but is not explained in the website). We can name it as "**if-MD**" for now.
   - **mineral-specimen.label-2** either has value "R" or no value. Label "R" actually stands for being marked in red colour in the browser, indicating  "The age displayed for this mineral is coming from a different, non-child locality". I suggest that we name it "**indirect_locality_date**", which is the same id used by RRUFF website in their html file. More important findings are:
     - Whenever it is labeled with "R", it is always labeled with one of the following labels in the label-3. (Reverse is not true) Although having label "R" and "G" seems unreasonable, such situations do exists, but very rare, and could be because of errors in the original data.
     - More importantly, I found out whenever it is labeled with "R", that row will contain a value of "dated locality", which is very straightforward: the date comes from different, non-child locality. **AND IT SEEMS THE REVERSE IS ALSO TRUE.** In summary:
       
       "R" 
       <==> "The age displayed for this mineral is coming from a different, non-child locality"
       <==> having a value in "dated locality", the value being the ID of the dated locality, and the record in the dated locality is always right above!
       
       Example: http://rruff.info/mineral_list/locality.php?mineral_name=Alunite
   - **mineral-specimen.label-3** could have one of the values below and I suggest rename it as "**date-info**"
     - "G" (green): This mineral is directly dated.
     - "B" (blue): This mineral is explicitely specified as having an age
     - "Y" (yellow): This mineral is using an age from a mineralization period.
     - "O" (orange): This mineral is using an age calculated from all data at the locality.
     - or no color: has no age

These suggested new names might not be final, but at least they carry some known information for now. We will need to communicate with RRUFF people to further verify and let them explain more about their conventions.
@zednis @kolans @xgmachina Thoughts?

**Comment by zednis** [comment]
> An initial suggestion is to call it "age-determination-process", but maybe a slightly different "age-determination-info" could be a more precise name, as a "process" could be a way bigger concept.

please elaborate

**Comment by zednis** [comment]
@zhonghao1014 can you provide an embedded example JSON in the comments (it does not need to be committed to the codebase, use [code blocks](https://help.github.com/articles/creating-and-highlighting-code-blocks/)) for what the proposed updates would look like?

**Comment by zhongh** [comment]
@zednis @xgmachina @kolans So, to elaborate for our discussion, a revised sample mapping/data structure could be like this below. A mineral observation contains information regarding the "mineral-species", "locality", and "mineral-specimen" which contains specific information regarding that specimen.
Under "mineral-specimen", the properties are "max-age", "min-age", "age-unit" (usually "Ma"), and an "age-determination-info" object which records information on how the age is determined. "age-determination-info" has three properties:
- "is-MD" = "MD" or null
- "indirect_locality_date" = "R" or null
- "date-info" = "G"/"B"/"Y"/"O" or null

Also, "age-determination-process" might tend to be a bigger object which also contains how the age is determined, the tools, the tech-flow, the determiner, the organization, the calculation methods .... So that why maybe we could use "age-determination-info" to imply that it only contains some info? Just a thought.

```
{  
  "mineral-observation":{  
    "properties":{  
      "locality":{  
        "properties":{
          (same)
        }
      },
      "mineral-species":{  
        "properties":{
          (same)
        }
      },
      "mineral-specimen":{  
        "properties":{  
          "max-age":{  
            "type":"double"
          },
          "min-age":{  
            "type":"double"
          },
          "age-unit":{  
            "type":"string"
          },
          "age-determination-info":{  
            "properties":{  
              "if-MD":{  
                "type":"string"
              },
              "indirect_locality_date":{  
                "type":"string"
              },
              "date-info":{  
                "type":"string"
              }
            }
          }
        }
      }
    }
  }
}
```

Suggestions?

**Comment by zednis** [comment]
I am not fond of any of the three suggested property names.
- `if-MD`: let's find out what this means from Josh.  Also, if it really only does have 2 possible values we should model it as a boolean property
- `indirect_locality_date`: This seems ok for now, but I would like verification from Josh on what this means, specifically for locality hierarchies.  As with MD, if there is only 2 values for this property I think we should model it as a boolean.
- `date-info`: age-source?

Also, for date-info/age-source where we have keys that are shown on the interface with descriptions I think the JSON should include both the key (or label) and the description.  This way the interface will not be

```
"date-info": {
  "label": "G",
  "description": "This mineral is directly dated"
}
```

**Comment by zhongh** [comment]
Sure! Next step I will do: schedule a quick call with Josh

**Comment by zhongh** [comment]
After communication with Josh, here is a summary on RRUFF labels and marking and suggested mapping:

# First 3 Fields in RRUFF Evolution Data:

## 1st Field (RED/GREEN and MD/null)

RED marking simply means that it does not qualify to be GREEN for that locality, but we can still list the mineral at that locality as some GREEN records can be passed up towards it.  For example, "lindgrenite" has a GREEN locality record at "Sierra Gorda District, Antofagasta Province, Antofagasta Region, Chile". If we click "Show Parent Localities", we can see that table shows that "[lindgrenite](http://rruff.info/mineral_list/locality.php?mineral_name=Lindgrenite)" is also listed, but marked RED, at 
"Antofagasta Province, Antofagasta Region, Chile",
"Antofagasta Region, Chile", and of course,
"Chile"
Unless we choose "Show Parent Localities", these RED-marked localities are not displayed in the RRUFF evolution data, and either way, as Josh put, they "should not be considered when counting the number of mineral localities".
[**RED/GREEN information in the first field is not provided from the data export. Need solution from RRUFF team.]

GREEN marking means at least one of the following conditions is met:
1) Mindat claims that this mineral occurs at this locality.
2) The locality containing the mineral has an age directly assigned to it.
3) This locality is displaying an age from a non-child locality."
E.g. "[lindgrenite](http://rruff.info/mineral_list/locality.php?mineral_name=Lindgrenite)" 

## 2nd Field (R/null)

A row has values in"Max Age" & "Min Age" when and only when it has a value in "Dated Locality ID". And field 2 shows label R if and only if "Dated Locality ID"≠"MinDat ID", which it simply means that the age is from another locality. 
R stands for "Remote" and more explicitly:
R: The age displayed for this mineral is coming from a different, non-child locality.

## 3rd Field (G/B/Y/O)

Same with the 3rd field of the associated dated locality. This is how the age is actually  determined.
G: This mineral is directly dated.
B: This mineral is explicitely specified as having an age. 
Y: This mineral is using an age from a mineralization period.
O: This mineral is using an age calculated from all data at the locality.

# Suggested Mapping Structure

A mineral observation contains information regarding the "mineral-species", "locality", and "mineral-specimen" which contains specific information regarding that specimen.
Under "mineral-specimen", the properties are "max-age", "min-age", "age-unit" (usually "Ma"), and an "age-determination-process" object which records information on how the age is determined. Like this

```
{  
  "mineral-observation":{  
    "properties":{  
      "locality":{  
        "properties":{  
          (same)
        }
      },
      "mineral-species":{  
        "properties":{  
          (same)
        }
      },
      "mineral-specimen":{  
        "properties":{  
          "max-age":{  
            "type":"double"
          },
          "min-age":{  
            "type":"double"
          },
          "age-unit":{  
            "type":"string"
          },
          "age-determination-process":{  
            "properties":{  
              "mindat-claims-occurence":{  
                "type":"boolean"
              },
              "remote-date":{  
                "properties":{  
                  "is-remote":{  
                    "type":"boolean"
                  },
                  "description":{  
                    "type":"string"
                  },
                  "dated-locality-label":{  
                    "type":"string"
                  },
                  "dated-locality-mindat-id":{  
                    "type":"string"
                  },
                  "dated-locality-mindat-url":{  
                    "type":"string"
                  }
                }
              },
              "date-info":{  
                "properties":{  
                  "label":{  
                    "type":"string"
                  },
                  "description":{  
                    "type":"string"
                  }
                }
              }
            }
          }
        }
      }
    }
  }
```

**Comment by zednis** [comment]
I am not sure why "remote-date" would have a child-field "is-remote"?  What would the following mean?

```
"remote-date": {  
  "is-remote": false,
  "description": "...",
  "dated-locality-label": "...",
  "dated-locality-mindat-id": "...",
  "dated-locality-mindat-url": "..."
  },
```

Also, what are examples of values for "dated-locality-label" and why would this information not be in the locality?

**Comment by zednis** [comment]
@zhonghao1014 Perhaps you should provide some examples of "age-determination-process"?  I am not sure how the current structure helps with search or faceted browsing.

edit - actually provide an example of a full record including locality info.  Pick a few examples with different scenarios so we can see how different scenarios map into the structure.

**Comment by zednis** [comment]
@zhonghao1014 I am going to close and not merge this pull request.  We need to create separate pull requests for each of the different updates/changes.

1 pull request for the min/max age facets
1 pull request for schema changes around locality
1 pull request for adding geo_point back into locality
1 pull request for schema changes around mineral dating information

We should also be making comments such as your description of what the different colors and codes mean in the issues and not the pull requests.  Pull request comments should be about the code construction, issue comments are about the general issue.

**Comment by zhongh** [comment]
Sure. Mixing multiples tasks in one pull request does go against sanity. How should we do it, since we already made some progress into this branch? @zednis 

Or I can just still finish up this branch and then create further branch to make changes. And also avoid this practice in the future?

I will provide examples of the "age-determination-process" object when we create a specialised branch for it after we close this.

**Comment by zhonghao1014** [comment]
I will provide examples of the "age-determination-process" object when we create a specialised branch for it after we close this.

**Comment by zednis** [comment]
@zhonghao1014 We'll just have to re-create the changes in the new branches.  After being closed the pull request and related commits will still be viewable.

Closing now.

**Pull request closed without merging**